service/worker/deployment/deployment_reachability.go

```diff
@@ -0,0 +1,195 @@
+	// The deployment is reachable by new and/or open workflows. The deployment cannot be
+	// decommissioned safely.
+	DEPLOYMENT_REACHABILITY_REACHABLE DeploymentReachability = 1
+	// The deployment is not reachable by new or open workflows, but might be still needed by
+	// Queries sent to closed workflows. The deployment can be decommissioned safely if user does
+	// not query closed workflows.
+	DEPLOYMENT_REACHABILITY_CLOSED_WORKFLOWS_ONLY DeploymentReachability = 2
+	// The deployment is not reachable by any workflow because all the workflows who needed this
+	// deployment went out of retention period. The deployment can be decommissioned safely.
+	DEPLOYMENT_REACHABILITY_UNREACHABLE DeploymentReachability = 3
+)
+*/
+
+func getDeploymentReachability(
+	ctx context.Context,
+	namespaceEntry *namespace.Namespace,
+	seriesName, buildID, currentBuildID string,
+	currentBuildIDValidTime time.Time,
+	cache reachabilityCache,
+) (enumspb.DeploymentReachability, time.Time, error) {
+	// 1a. Reachable by new unpinned workflows
+	if buildID == currentBuildID { // add if buildID is ramping, once we have ramp
+		return enumspb.DEPLOYMENT_REACHABILITY_REACHABLE, currentBuildIDValidTime, nil
+	}
+
+	// 2a. Reachable by open pinned workflows
+	countRequest := visibility_manager.CountWorkflowExecutionsRequest{
+		NamespaceID: namespaceEntry.ID(),
+		Namespace:   namespaceEntry.Name(),
+		Query:       makeDeploymentQuery(seriesName, buildID, true),
+	}
+	exists, lastUpdateTime, err := cache.Get(ctx, countRequest, true)
+	if err != nil {
+		return enumspb.DEPLOYMENT_REACHABILITY_UNSPECIFIED, time.Time{}, err
+	}
+	if exists {
+		return enumspb.DEPLOYMENT_REACHABILITY_REACHABLE, lastUpdateTime, nil
+	}
+
+	// 3. Reachable by closed pinned workflows
+	countRequest = visibility_manager.CountWorkflowExecutionsRequest{
+		NamespaceID: namespaceEntry.ID(),
+		Namespace:   namespaceEntry.Name(),
+		Query:       makeDeploymentQuery(seriesName, buildID, false),
+	}
+	exists, lastUpdateTime, err = cache.Get(ctx, countRequest, false)
+	if err != nil {
+		return enumspb.DEPLOYMENT_REACHABILITY_UNSPECIFIED, time.Time{}, err
+	}
+	if exists {
+		return enumspb.DEPLOYMENT_REACHABILITY_CLOSED_WORKFLOWS_ONLY, lastUpdateTime, nil
+	}
+
+	// 4. Unreachable
+	return enumspb.DEPLOYMENT_REACHABILITY_UNREACHABLE, time.Now(), nil
+}
+
+func makeDeploymentQuery(seriesName, buildID string, open bool) string {
+	var statusFilter string
+	deploymentFilter := sqlparser.String(sqlparser.NewStrVal([]byte(
+		worker_versioning.ReachabilityBuildIdSearchAttribute(enumspb.VERSIONING_BEHAVIOR_PINNED, &deploymentpb.Deployment{
+			SeriesName: seriesName,
+			BuildId:    buildID,
+		}),
+	)))
+	if open {
+		statusFilter = `= "Running"`
+	} else {
+		statusFilter = `!= "Running"`
+	}
+	// todo (carly): handle null / unversioned
+	return fmt.Sprintf("%s = %s AND %s %s", searchattribute.BuildIds, deploymentFilter, searchattribute.ExecutionStatus, statusFilter)
```

Comment:
maybe missing a second =

Reply:
The "statusFilter" has either = or != in it already, so it's there.
Might be more readable if I made the "deploymentFilter" also come with its own = for symmetry though